### PR TITLE
Fix S2259 FP: Support TypeOf operation

### LINKED ISSUE #6157: Fix S2259 FP: Support TypeOf operation
Learn a NotNull constraint from `typeof` operation.

```
var type = typeof(NullPointerDereference);
if(type==null)
{
    // Unreachable
}
```
to fix
```
  var other = typeof(object);
  var type = typeof(NullPointerDereference);
  if (other.IsAssignableFrom(type))
  {
      type.ToString();
  }
  else
  {
      type.ToString();    // Compliant, unreachable for .NET, no annotation under .NET Fx
  }
```

From Peach:
https://peach.sonarsource.com/project/issues?issues=AYOQ3tRxWAUSpAKvfrGP&open=AYOQ3tRxWAUSpAKvfrGP&id=aspnetcore

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #6157

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs [hunk 1]
```diff
@@ -540,6 +540,20 @@ void ShortCircuit_Reachable()
                 o.ToString();   // Compliant for the same reason
             }
         }
+
+        public void Type_IsAssignableFrom()
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs:544]: This is the only test in NullPointerDerefernce that is about `typeof`. Add one more test case here, that comes without any "IsAssignable" complications, like e.g.
```cs
Type type = null;
type = typeof(object);
type.ToString(); // Compliant
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.NetCore.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.NetCore.cs:1]: I already created some NetCore specific tests in `NullPointerDereference.CSharp8.cs` (e.g. tests in `ThrowHelper` and `NotNullWhenTests`). The reasoning was, that NRT was introduced along with C#8 and with it the annotation attributes. So the annotation attributes are tied to .Net Core 5 and .NetStandard 2.1 and C#8 is the oldest official supported language version for these runtimes.
I think it is good to have `NullPointerDereference.NetFx.cs` but `NullPointerDereference.NetCore.cs` looks like a duplicate to `NullPointerDereference.CSharp8.cs` to me.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
NullPointerDereference.NetCore.cs may not be needed as there is already a fitting test file.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs [hunk 3]
```diff
@@ -540,6 +540,20 @@ void ShortCircuit_Reachable()
...
+        {
+            var other = typeof(object);
+            var type = typeof(NullPointerDereference);
+            if (other.IsAssignableFrom(type))
+            {
+                type.ToString();
+            }
+            else
+            {
+                type.ToString();    // Compliant, unreachable for .NET, no annotation under .NET Fx
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs:554]: That's not the root cause

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.NetCore.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.NetCore.cs:1]: The problem is that `CSharp8.cs` is missused to test `NetCore`-only stuff.

C# 8 is language version and needs to be tested under NetFx too.
I've moved the test cases from there to `.NetCore.cs`, independent on language version. And made the C# 8 UT run under both TFMs

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs]: But it isn't unreachable for .Net. `IsAssignableFrom` may return `false`.

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM. One code comment is still misleading if I'm not mistaken.

**Pull request merged**